src/java.base/share/classes/java/nio/Direct-X-Buffer.java.template

```diff
@@ -509,8 +509,55 @@ class Direct$Type$Buffer$RW$$BO$
                                             offset, segment);
     }
 
-#end[char]
+    public $Type$Buffer put(int index, CharSequence csq, int start, int end) {
+#if[rw]
+        Objects.checkFromToIndex(start, end, csq.length());
+        Objects.checkIndex(index, limit());
+        if (end - start > limit() - index)
+            throw new BufferOverflowException();
+
+        //
+        // Cannot delegate to put(int,CharBuffer,int,int) until after the check
+        // for BufferOverflowException as that method would instead throw an
+        // IndexOutOfBoundsException for the same condition
+        //
+        if (csq instanceof CharBuffer cb)
+            return put(index, cb, start, end - start);
+
+        if (csq instanceof String ||
+            csq instanceof StringBuilder ||
+            csq instanceof StringBuffer) {
+
+            char[] buf = new char[Math.min(APPEND_BUF_SIZE, end - start)];
+            while (start < end) {
+                int count = end - start;
+                if (count > buf.length)
+                    count = buf.length;
+
+                if (csq instanceof String str) {
+                    str.getChars(start, start + count, buf, 0);
+                } else if (csq instanceof StringBuilder sb) {
+                    sb.getChars(start, start + count, buf, 0);
+                } else if (csq instanceof StringBuffer sb) {
+                    sb.getChars(start, start + count, buf, 0);
```

Comment:
Can the allocation and double copy be avoided?
But I guess the combinatorics of latin1/utf16 string encoding and the memory and byte order make it difficult.

Reply:
Back when I was originally developing this change I could not see any better way to do this part.

Reply:
> Can the allocation and double copy be avoided?

The performance improvement in the heap and direct implementations versus the `X-Buffer` base implementation comes from using `getChars` to process a chunk of `char`s at once instead of doing them one at a time with `charAt`. The heap implementation can avoid the allocation and double copy by passing in its array to `getChars` directly, but in the direct implementation this is not possible. I could not identify a better way for the direct version.